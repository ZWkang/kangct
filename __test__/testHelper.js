// import React from 'react';
import { mount, shallow, render } from 'enzyme';

function setUp(type, ...arg) {
  const [Component, props = {}, children = ''] = arg;
  if (~['mount', 'shallow', 'render'].indexOf(type)) {
    return global[type](<Component {...props}>{children}</Component>);
  }
}

export function AutoMath(Component, props = {}, children = '') {
  const wrapper = setUp('render', Component, props, children);
  expect(wrapper).toMatchSnapshot();
}

export function ComponentRenderWithoutError(Component, props = {}, children = '') {
  const wrapper = setUp('mount', Component, props, children);
  expect(() => {
    wrapper.setProps({});
    wrapper.unmount();
  }).not.toThrow();
}

export function childrenShouldBeProps(Component, children) {
  const wrapper = setUp('mount', Component, {}, children);

  expect(wrapper.get(0).props.children).toEqual(children);
}

export function setup(Component, props) {
  const wrapper = setUp(Component, props, null);
  return {
    wrapper,
    props
  };
}
