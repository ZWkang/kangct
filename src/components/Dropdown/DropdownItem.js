import React, { Component } from 'react'
import styled from 'styled-components'

const Item = styled.li`
    display: flex;
    height: 36px;
    padding: 0 16px;
    line-height: 36px;
    color: #333;
    cursor: pointer;
    :hover {
        background: rgba(66,133,244,.08);
    }
`

const ItemTitle = styled.div`
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
`

const HrLine = styled.div`
    height: 1px;
    overflow: hidden;
    background-color: #e7e7e7;
    line-height: 0;

`

class DropdownItem extends Component {
    constructor(props) {
        super(props)
    }
    handleItemClick = () => {
        const { onClick } = this.props
        onClick && onClick()
    }

    render() {
        const { title } = this.props
        if(title === '-') {
            return (
                <HrLine/>
            )
        }
        return (
            <Item onClick = {this.handleItemClick}>
                <ItemTitle>
                    {title}
                </ItemTitle>
            </Item>
        )
    }
}
export default DropdownItem