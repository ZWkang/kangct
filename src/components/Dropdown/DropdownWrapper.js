import React from 'react'
import styled from 'styled-components'

const UlContainer = styled.ul`
    box-sizing: border-box;
    width: 100%;
    min-width: 140px;
    margin-top: 2px!important;
    margin-bottom: 2px;
    /* padding: 4px 0!important; */
    border: 1px solid #d8d8d8;
    list-style: none;
    font-size: 14px;
    text-align: left;
    background-color: #fff;
    background-clip: padding-box;
    box-shadow: 0 8px 12px rgba(0,0,0,.175);
`



export {
    UlContainer
}