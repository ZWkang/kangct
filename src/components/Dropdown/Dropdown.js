import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import DropdownItem from './DropdownItem';
import Rect from '../Rect';
import { UlContainer } from './DropdownWrapper';
import enhanceWithClickOutside from 'react-click-outside';

const Temp = styled.div`
  position: relative;
  text-align: left;
  padding: 0 6px;
  cursor: pointer;
`;
function noop() {}
function defaultRenderTitle({ title }) {
  return <span>{title}</span>;
}
class Dropdown extends Component {
  static defaultProps = {
    show: false,
    list: [],
    onClick: noop
  };
  constructor() {
    super();
    this.dropdown = React.createRef();
    this.state = {
      dropdownShow: false
    };
  }
  handleTitleClick = () => {
    const { dropdownShow } = this.state;
    this.setState({
      dropdownShow: !dropdownShow
    });
  };
  registerEvents = () => {
    let { trigger } = this.props;
    if (typeof trigger === 'string') {
      trigger = [trigger];
    }
    trigger.map((triggerItem) => {
      this.dropdown.current.addEventListener(triggerItem, this.triggerEvent, false);
    });
  };
  triggerEvent = (e) => {
    const { dropdownShow } = this.state;
    e && e.preventDefault();
    this.setState({
      dropdownShow: !dropdownShow
    });
  };
  componentDidMount() {
    this.registerEvents();
  }
  handleClickOutside = () => {
    if (this.state.dropdownShow) {
      this.setState({
        dropdownShow: false
      });
    }
  };
  renderTitle = () => {
    const { renderTitle, title } = this.props;
    if (typeof renderTitle === 'function') {
      return renderTitle({
        title: title
      });
    }
    return defaultRenderTitle({
      title: title
    });
  };
  handleItemClick = () => {};
  render() {
    const { title } = this.props;
    const { dropdownShow } = this.state;
    return (
      <Temp ref={this.dropdown}>
        {this.renderTitle()}
        <Protal>
          <Rect target={this.dropdown} show={dropdownShow}>
            {({ left, top, height }) => {
              return (
                <div
                  style={{
                    position: 'absolute',
                    top: top + height + 'px',
                    left: left + 'px',
                    display: 'block',
                    display: dropdownShow ? '' : 'none'
                  }}
                >
                  <UlContainer style={{ padding: '0' }}>
                    {['test', 'xxx'].map((item) => {
                      return <DropdownItem title={item} />;
                    })}
                  </UlContainer>
                </div>
              );
            }}
          </Rect>
        </Protal>
      </Temp>
    );
  }
}

/**
 * @components
 */
export default enhanceWithClickOutside(Dropdown);
