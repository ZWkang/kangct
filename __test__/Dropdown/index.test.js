import 'jest-styled-components';
import Dropdown from '@app/Dropdown/index';
import DropdownItem from '@app/Dropdown/DropdownItem';
import { ComponentRenderWithoutError, AutoMath } from '../testHelper.js';

jest.mock('../../src/components/Rect/index.js', () => (props) => {
  const { children } = props;

  return children({
    left: 750,
    top: 320,
    height: 300
  });
});
jest.mock('../../src/components/Protal/index.js', () => (props) => {
  return <div>{props.children}</div>;
});

// jest.mock('installComponent', () => ({
//   InstallCom: 'mockInstallCom'
// }));

describe('Dropdown', () => {
  describe('Dropdown rendering', () => {
    // <Dropdown
    //   list={['item1', 'item2']}
    //   title={'click me please'}
    //   trigger={['click']}
    //   onClick={(item, e) => {
    //     console.log(item);
    //   }}
    // />;
    it('without error', () => {
      ComponentRenderWithoutError(
        Dropdown,
        {
          list: ['item'],
          title: 'test',
          trigger: ['click'],
          onClick: jest.fn()
        },
        null
      );
    });
    it('auto snapshot', () => {
      const wrapper = shallow(<Dropdown></Dropdown>, {
        list: ['item'],
        title: 'test',
        trigger: ['click'],
        onClick: jest.fn()
      });
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('Dropdown Interactions', () => {
    it('Dropdown item click', () => {
      const mockClickFn = jest.fn();
      const DropdownProps = {
        list: ['item'],
        title: 'test',
        trigger: ['click'],
        onClick: mockClickFn
      };
      const wrapper = mount(<Dropdown {...DropdownProps}></Dropdown>);

      wrapper
        .find(DropdownItem)
        .at(0)
        .simulate('click');
      expect(mockClickFn).toBeCalled();
      expect(mockClickFn).toHaveBeenCalledTimes(1);
    });
  });
  describe('Dropdown Interactions', () => {});
});
