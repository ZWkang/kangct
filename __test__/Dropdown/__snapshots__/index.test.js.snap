// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Dropdown rendering auto snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <clickOutside(Dropdown) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "list": Array [],
      "onClick": [Function],
      "show": false,
    },
    "ref": [Function],
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "list": Array [],
        "onClick": [Function],
        "show": false,
      },
      "ref": [Function],
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "list": Array [
      "item",
    ],
    "onClick": [MockFunction],
    "title": "test",
    "trigger": Array [
      "click",
    ],
  },
  Symbol(enzyme.__childContext__): null,
}
`;
